Instructor pages

alex-shen1 opened the pull request:
# Status
Done

# What I did

Instructor info page. Currently just have skeleton for it in `dev`, and this PR fleshes out content/styling.

## Issues addressed

Closes #180 

## Screenshots

Old site:

![image](https://user-images.githubusercontent.com/55100084/98064029-5c856e80-1e1f-11eb-9f1b-2c1fc8cd6041.png)

New site:
![image](https://user-images.githubusercontent.com/45245651/98837499-057c2c80-2411-11eb-83d7-8610a03f2edd.png)

jackliu612 reviewed (commented):
Looks pretty good overall, Just some nitpicks and performance comments. Personally didn't look too closely at css changes as I'm not familiar so might want another reviewer to look over. One thing I'm not sure is that it says legacy_db/mysql2sqlite, manage.py, precommit were modified not sure what that's about.

Bartleby2718 commented on tcf_website/views/browse.py:
```diff
@@ -148,14 +148,70 @@ def course_instructor(request, course_id, instructor_id):
+    """
+    if num is not None:
+        return round(num, 2)
+    return '—'
```
+1 for `'\u2014'` but I don't think pylint recommends `if ... return ... else ... return ...`. I remember fixing a build error because of that.

alex-shen1 replied:
Done — not sure if it actually works because builds can't work on my machine anymore thanks to DB update, and I don't feel like reverting and then re-reverting just to see if an em dash shows up.

alex-shen1 replied:
Also, can confirm that pylint throws an error if you try to add else.

alex-shen1 commented on tcf_website/models/models.py:
```diff
@@ -213,6 +207,21 @@ def average_difficulty(self):
+            Section.objects.filter(
+                instructors=self).distinct().values_list(
+                'course_id', flat=True))
+        # TODO: remove old courses that haven't been taught in 5+ years;
```
I'm not sure what I was thinking here....I think the idea was that 5+ years is the scope of relevancy for a review (because things probably change), but tCF 1 just keeps all review data. I think it's pretty safe to leave out anything with a 3-digit course number because I highly doubt UVA will ever go back to that, and manually figuring out what got renamed is probably a wasted effort considering how few reviews are likely from that time anyways.

I'll just remove that comment line, though.

alex-shen1 commented on tcf_website/views/browse.py:
```diff
@@ -148,14 +148,70 @@ def course_instructor(request, course_id, instructor_id):
+
+    for course in courses:
+        course_subdept = course.subdepartment.name
+        course_data = {
```
yeah this was autoformat's doing, i'll fix it

alex-shen1 replied:
Actually, autopep8 automatically brings it back to this wack status.....so any time someone runs `precommit` it'll revert the change.

julesfll replied:
gonna have to do an autopep8 ignore

alex-shen1 replied:
do you know how to do that? looked into it briefly and couldn't find anything useful

julesfll replied:
idk

alex-shen1 commented on tcf_website/views/browse.py:
```diff
@@ -148,14 +148,70 @@ def course_instructor(request, course_id, instructor_id):
+            map(
+                lambda subdept_id: [
+                    subdept_id,
+                    Subdepartment.objects.get(
```
I think since each professor is likely to only teach in a few subdepartments, this'll only run at most 3-4 times, so it's not urgent to fix, but I'll throw it on the backlog in case we ever wanna take a crack at it.

alex-shen1 replied:
Issue #210

julesfll commented on tcf_website/views/browse.py:
```diff
@@ -148,14 +148,70 @@ def course_instructor(request, course_id, instructor_id):
+                    instructor, course)), 'avg_difficulty': safe_round(
+                Instructor.average_difficulty_for_course(
+                    instructor, course)), 'avg_hours': (
+                        lambda x: safe_round(x) if x is not None else '\u2014')(
```
Could we just delete this part since `safe_round` already does error checking
```suggestion
                        lambda x: safe_round(x)) (
```

alex-shen1 replied:
Originally there was separate code because I wanted to truncate hours and not round to 2 decimal points, but then the decision to keep precision on hours as well rendered this invalid. Good catch, but I guess since we're removing hours entirely it doesn't matter anymore.

alex-shen1 commented:
This is kinda outside the scope of the PR, but I tweaked the course-instructor page a little bit too. Firstly, I changed the size/position of course title and instructor name to be more in line with tCF1 (even though it's the instructor's page for the class, I think the class title should still be on top because it's the "bigger" category). However, simply switching it like this made it look like this when you narrowed the window:
![image](https://user-images.githubusercontent.com/55100084/99162856-8157cc80-26d0-11eb-98de-d383b0bc273f.png)
It didn't seem coherent to me that the mnemonic would be continuing on the bottom when people would naturally read left-to-right, so to avoid a "don't dead open inside" situation, I fixed the wrapping so it looks like this:
![image](https://user-images.githubusercontent.com/55100084/99162888-f88d6080-26d0-11eb-8991-68ed527a721d.png)
![image](https://user-images.githubusercontent.com/55100084/99162891-017e3200-26d1-11eb-8d8a-efad7cffad0b.png)

Styling isn't really my "thing" though, so let me know if it looks bad.

j-alicia-long commented on tcf_website/static/base/tcf-style.css:
```diff
@@ -19,10 +19,15 @@ h2,
+.h6 {
   /* Default heading font: Lato (others are fallbacks) */
   font-family: 'Lato', 'Helvetica Neue', Helvetica, Arial, sans-serif;
-  font-weight: 700;
```
Are we making the font-weight thinner?

julesfll replied:
yeah I talked to you abt that and you said it looked better thinner
I agree
it is global but I think it looks good where I checked

j-alicia-long replied:
gotcha yeah I just forgot which branch you did it on

j-alicia-long reviewed (commented):
Looks good overall! Thanks for drafting this all up quickly.

alex-shen1 commented on tcf_website/models/models.py:
```diff
@@ -222,6 +216,22 @@ def average_difficulty(self):
+                'course_id', flat=True))
+
+        # <1000 course IDs are from super old classes...
+        # should we bother keeping the data at that point?
```
I think so? It's not immediately obvious from looking at the code why it's done like that, it gives information that might be helpful in the future (i.e. maybe a reminder of something to pick back up on later), and it doesn't really hurt anything.
